### PR TITLE
fix(timeouts): Amp 34805

### PR DESCRIPTION
**Pull request by ajhorst**
### Summary

Simplify how timeouts are handled for event logging and solve a few issues with it

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-Node/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?: No

### CODE REVIEW

**Comment by ajhorst** [comment]
@kelvin-lu this is my first pass at fixing the timeout problem after our conversation yesterday. The TLDR is that I took the timeout logic from the queue and instead made each HTTP request time itself out. This has a few ramifications:
- the queue is simpler and no longer has a few edge case bugs (the timeout was never called on the first promise in the queue, for example)
- The timeout timer for each promise now starts when the promise request begins, instead of when it's put into the queue. I suspect that old behavior is why people have been complaining a lot about timeouts lately. 
- when a promise times out, it rejects and immediately starts the next one. There was the possibility of some delay between them in the previous implementation.

If it's important for the queue to have the timeout logic and not the promise implementation, then I also did some reading on how we could do that while also keeping the behavior of the promise rejecting on timeout (see ttps://medium.com/@masnun/creating-cancellable-promises-33bf4b9da39c). 

Let me know what you think with this approach!

**Review by kelvin-lu — COMMENTED** [review]
I think this makes sense! Wanted to raise that this changes the nature of the timing of the request timeout though.

Before, the timeout starts when the promise was *added* to the queue - so it could timeout if it was queued up for a very long time. Now, the timeout starts when the request starts, so the timeout only happens if the request takes too long. 

Maybe this is better though 🤔

File: packages/node/src/transports/http.ts
Comment (ajhorst) [reply packages/node/src/transports/http.ts:82]: ah, yes, that's a good catch.

**Comment by ajhorst** [comment]
@kelvin-lu Yeah, this does change the nature of the timeouts. I have a good idea of how to change this to keep the general behavior of rejecting the promise on timeout, but still starting the timer when the new promises are enqueued. I suppose it's a sort of product decision, then, which timeout approach is better. I defer to you on that

**Comment by kelvin-lu** [comment]
@ajhorst I'm scratching my head to remember if there was a reason the timeout is on the start of the queue and not when it starts the request, but I **think** this makes more sense.

It's a question of what this means for optioning - is the the maximum time of a singular request, or an estimate of the maximum time the promise `flush()` should resolve in?

**Comment by ajhorst** [comment]
@kelvin-lu I'm wondering if there are cases in the current code where 
1) a log request is queued, and the timer starts
2) other parts of the client code take a while to execute (maybe some heavy synchronous file IO or network requests)
3) the queue times out, even though very little of the 10 seconds was spent in the queue itself

does that sound plausible? I'm trying to think of situations that could case the big spike in timeouts here.

File: packages/node/src/transports/http.ts
Comment (ajhorst) [reply packages/node/src/transports/http.ts]: @kelvin-lu added `req.destory()` to the timeout (apparently req.abort() is deprecated)

**Comment by kelvin-lu** [comment]
@ajhorst I think so! They could also be `flush`-ing super often, and causing a pile-up of network requests (or perhaps the network requests are being resolved really slowly?) I think though that this control flow makes more sense though.

File: packages/utils/src/queue.ts
Comment (kelvin-lu) [inline packages/utils/src/queue.ts:17]: I think semver purists would call this a breaking change 🤔 @jooohhn strong thoughts here? we could keep this and mention that it's deprecated, though since no one has listed our utils as a dependency the chance that this would upset someone is low

File: packages/node/src/transports/http.ts
Comment (kelvin-lu) [inline packages/node/src/transports/http.ts:111]: small ask - since we now have more control of the response, maybe we could include a `Body` that includes metadata on how long the request waited before timing out in ms, or when the network request started? Non-blocking but would prefer to have a TODO/follow-up if not addressed currently.

This might also help our users debug what exactly is going on,

**Review by kelvin-lu — COMMENTED** [review]
Overall LGTM, I think the next ask is if it's possible to add a unit test for this - if nock allows for us to force a request to wait a certain amt of time before resolution

File: packages/utils/src/queue.ts
Comment (kelvin-lu) [reply packages/utils/src/queue.ts:17]: I think we're about ripe to have `2.0.0` in collecting a bunch of deprecated code pieces though - maybe we should also discuss a strategy about going about that 😄

Comment (ajhorst) [reply packages/utils/src/queue.ts:17]: That raises an interesting question- do we consider all the code in this repo to be public-facing? I wouldn't have thought that anyone would go digging into the internals of our logic for queueing and sending requests.

File: packages/node/src/transports/http.ts
Comment (ajhorst) [inline packages/node/src/transports/http.ts:49]: Tweaked the signature of `sendPayload` to accept a timeout miliseconds number and pass that down to _sendWithModule. This makes it possible to test with nock,  plus there's a ticket out to make the timeout period an optional parameter to the client, and this helps pave the way for that change.

**Review by kelvin-lu — APPROVED** [review]

File: packages/node/test/http.test.ts
Comment (kelvin-lu) [reply packages/node/test/http.test.ts]: I managed to comment on already outdated code, please ignore

File: packages/node/src/transports/http.ts
Comment (ajhorst) [reply packages/node/src/transports/http.ts:111]: I'll make a ticket to follow up on this

**Pull request merged**
